### PR TITLE
Atomic improvements

### PR DESCRIPTION
**Pull request by ChrisJefferson**
This is some improvements to the atomic list code.

* Make COMPARE_AND_SWAP work for variable length lists
* Add ATOMIC_BIND and ATOMIC_UNBIND, which act like COMPARE_AND_SWAP but for binding and unbinding.
* Add CHECKED_GET, which atomicly checks if a location is bound, and if so get the value.

### CODE REVIEW

File: lib/test.gi [hunk 1]
```diff
@@ -371,6 +371,14 @@ InstallGlobalFunction("Test", function(arg)
            end,
            subsWindowsLineBreaks := true,
          );
+
+  if IsBound(HPCGAP) then
```
Comment (fingolfin) [inline lib/test.gi]: Notes on the first commit message: "Disable the ... method" -> "Disable the ... method for HPC-GAP"

"sometimes add" -> "sometimes adds"

File: lib/hpc/thread1.g [hunk 2]
```diff
@@ -168,6 +175,43 @@ BIND_GLOBAL("CopyToRegion", {x,y} -> x);
...
+BIND_GLOBAL("CHECKED_GET_ATOMIC_LIST", function(l, pos)
+    if IsBound(l[pos]) then
+        return [true, l[pos]];
+    else
+        return [false];
+    fi;
+end);
+
+BIND_GLOBAL("COMPARE_AND_SWAP", function(l, pos, old, new)
+    if not(IsBound(l[pos]) and IS_IDENTICAL_OBJ(l[pos], old) ) then
```
Comment (fingolfin) [inline lib/hpc/thread1.g]: add space between `not` and `(`.

Also remove the space before the last `)`; or else, add a space after the first `(`.

Comment (fingolfin) [reply lib/hpc/thread1.g]: You could also switch the if and the else bodies, and get rid of the `not` (but leaving it as it is of course is also fine)

**Comment by ChrisJefferson** [comment]
I have made some improvements, and at least added documentation in the comments.

File: src/hpc/aobjects.c [hunk 3]
```diff
@@ -367,25 +411,33 @@ static Obj FuncSET_ATOMIC_LIST(Obj self, Obj list, Obj index, Obj value)
...
+    UInt n = INT_INTOBJ(index);
+
+    switch (TNUM_OBJ(list)) {
     case T_FIXALIST:
     case T_APOSOBJ:
       break;
+    case T_ALIST:
+        return AtomicCompareSwapAList(list, n, old, new);
     default:
       ArgumentError("COMPARE_AND_SWAP: First argument must be a fixed atomic list");
```
Comment (fingolfin) [inline src/hpc/aobjects.c]: Isn't that error incorrect now?

File: src/hpc/aobjects.c [hunk 4]
```diff
@@ -347,6 +347,50 @@ static Obj FuncGET_ATOMIC_LIST(Obj self, Obj list, Obj index)
...
+// could be unbound before the actual reading is performed.
+static Obj FuncCHECKED_GET_ATOMIC_LIST(Obj self, Obj list, Obj index)
+{
+    UInt        n;
+    UInt        len;
+    AtomicObj * addr;
+    Obj         val;
+    if (TNUM_OBJ(list) != T_ALIST && TNUM_OBJ(list) != T_FIXALIST)
+        ArgumentError(
+            "GET_ATOMIC_LIST: First argument must be an atomic list");
```
Comment (fingolfin) [inline src/hpc/aobjects.c]: `GET_ATOMIC_LIST` doesn't match the function name.

File: src/hpc/aobjects.c [hunk 5]
```diff
@@ -347,6 +347,50 @@ static Obj FuncGET_ATOMIC_LIST(Obj self, Obj list, Obj index)
   return addr[n+1].obj;
 }
 
+// Returns a list which contains one or two values. The first element
+// is the value of IsBound(list[index]). If list[index] is bound, the
+// list has a secnod element, which is the value of list[index].
```
Comment (fingolfin) [inline src/hpc/aobjects.c]: Typo: "secnod"

File: lib/hpc/thread1.g [hunk 6]
```diff
@@ -168,6 +175,43 @@ BIND_GLOBAL("CopyToRegion", {x,y} -> x);
...
+    if IsBound(l[pos]) then
+        return false;
+    else
+        l[pos] := new;
+        return true;
+    fi;
+end);
+
+BIND_GLOBAL("ATOMIC_UNBIND", function(l, pos, old)
+    if not (IsBound(l[pos]) and IS_IDENTICAL_OBJ(l[pos], old)) then
```
Comment (fingolfin) [inline lib/hpc/thread1.g]: How about getting rid of the `not` and swapping the if/else cases? Just as in `COMPARE_AND_SWAP`

**Comment by ChrisJefferson** [comment]
I have adjusted `CHECKED_GET_ATOMIC_LIST` to take an extra argument, which is the value to be returned if the value is not bound.

Once this is merged, I will write a `CheckedGet(list, index, default)`, which will be implemented for lists, atomic lists, and weak pointer lists, to provide a unified interface.

**Pull request merged**